src/map/core.rs

```diff
@@ -0,0 +1,409 @@
+use std::vec::Vec;
+
+use hashbrown::raw::RawTable;
+
+use std::cmp;
+use std::fmt;
+use std::mem::replace;
+use std::ops::RangeFull;
+use std::vec::Drain;
+
+use equivalent::Equivalent;
+use util::enumerate;
+use {Bucket, Entries, HashValue};
+
+/// Core of the map that does not depend on S
+pub(crate) struct IndexMapCore<K, V> {
+    /// indices mapping from the entry hash to its index.
+    indices: RawTable<usize>,
+    /// entries is a dense vec of entries in their order.
+    entries: Vec<Bucket<K, V>>,
+}
+
+#[inline(always)]
+fn get_hash<K, V>(entries: &[Bucket<K, V>]) -> impl Fn(&usize) -> u64 + '_ {
+    move |&i| entries[i].hash.get()
+}
+
+impl<K, V> Clone for IndexMapCore<K, V>
+where
+    K: Clone,
+    V: Clone,
+{
+    fn clone(&self) -> Self {
+        let indices = self.indices.clone();
+        let mut entries = Vec::with_capacity(indices.capacity());
+        entries.clone_from(&self.entries);
+        IndexMapCore { indices, entries }
+    }
+
+    fn clone_from(&mut self, other: &Self) {
+        let hasher = get_hash(&other.entries);
+        self.indices.clone_from_with_hasher(&other.indices, hasher);
+        if self.entries.capacity() < other.entries.len() {
+            // If we must resize, match the indices capacity
+            self.reserve_entries();
+        }
+        self.entries.clone_from(&other.entries);
+    }
+}
+
+impl<K, V> fmt::Debug for IndexMapCore<K, V>
+where
+    K: fmt::Debug,
+    V: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("IndexMapCore")
+            .field("indices", &raw::DebugIndices(&self.indices))
+            .field("entries", &self.entries)
+            .finish()
+    }
+}
+
+impl<K, V> Entries for IndexMapCore<K, V> {
+    type Entry = Bucket<K, V>;
+
+    #[inline]
+    fn into_entries(self) -> Vec<Self::Entry> {
+        self.entries
+    }
+
+    #[inline]
+    fn as_entries(&self) -> &[Self::Entry] {
+        &self.entries
+    }
+
+    #[inline]
+    fn as_entries_mut(&mut self) -> &mut [Self::Entry] {
+        &mut self.entries
+    }
+
+    fn with_entries<F>(&mut self, f: F)
+    where
+        F: FnOnce(&mut [Self::Entry]),
+    {
+        f(&mut self.entries);
+        self.rebuild_hash_table();
+    }
+}
+
+impl<K, V> IndexMapCore<K, V> {
+    #[inline]
+    pub(crate) fn new() -> Self {
+        IndexMapCore {
+            indices: RawTable::new(),
+            entries: Vec::new(),
+        }
+    }
+
+    #[inline]
+    pub(crate) fn with_capacity(n: usize) -> Self {
+        IndexMapCore {
+            indices: RawTable::with_capacity(n),
+            entries: Vec::with_capacity(n),
+        }
+    }
+
+    #[inline]
+    pub(crate) fn len(&self) -> usize {
+        self.indices.len()
+    }
+
+    #[inline]
+    pub(crate) fn capacity(&self) -> usize {
+        cmp::min(self.indices.capacity(), self.entries.capacity())
+    }
+
+    pub(crate) fn clear(&mut self) {
+        self.indices.clear_no_drop();
```

Comment:
Is there a particular reason you use `clear_no_drop` over `clear` here? The latter already optimizes for the case when the type doesn't need dropping: https://docs.rs/hashbrown/0.8.0/src/hashbrown/raw/mod.rs.html#612

Reply:
I suppose that's an overly-eager optimization on my part. Would it help your griddle port to use plain `clear`?

Reply:
It would — `clear_no_drop` is tricky in griddle since it's not clear how you clear the iterator of "leftover" elements without also dropping the backing storage for that iterator.

Reply:
Done.